### PR TITLE
[AERIE-1878] Add Hasura `constraintViolation` action

### PR DESCRIPTION
**Pull request by skovati**
* **Tickets addressed:** AERIE-1878
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

### No longer a breaking change, backwards compatibility maintained

Instead of constraint violations being included with the `results` of the `simulate` query, they are now accessed through a separate API call `constraintViolations`. Implemented so that querying `constraintViolations` before a simulation has been run will not start the simulation and instead return an empty JSON object. This was done in anticipation of future work that will *only* run the simulation on the `simulate` query.

This was implemented by adding a getter method to the `SimulationService` interface that allow the `getViolations` method in `GetSimulationResultsAction` to be called without passing simulation results, thus allowing us to query for constraint violations irrespective of simulation status. This getter logic was necessary so that while `UncachedSimulationService` will just run the simulation regardless, `CachedSimulationService` will return an empty `Map` if uncached.

## Verification

Tested with the banananation model with one PeelBanana activity, and the following constraint:
```TypeScript
export default (): Constraint =>
    Real.Resource("/fruit").equal(4)
```
which is violated when `peelDirection` is "fromStem"

### Old Behavior
Query of:
```GraphQL
query OldLameQuery {
  simulate(planId: 1) {
    status
    results
    reason
  }
}
```
returns:
```JSON
{
  "data": {
    "simulate": {
      "status": "complete",
      "results": {
        "unfinishedActivities": {},
        "start": "2022-020T12:12:12.000000",
        "activities": {
          "1": {
            "parent": null,
            "arguments": {
              "peelDirection": "fromStem"
            },
            "children": [],
            "computedAttributes": "UNIT",
            "type": "PeelBanana",
            "duration": 0,
            "startTimestamp": "2022-020T15:02:31.354000"
          }
        },
        "constraints": {
          "model/no_mashed": [
            {
              "windows": [
                {
                  "start": 10219354000,
                  "end": 86400000000
                }
              ],
              "associations": {
                "resourceIds": [
                  "/fruit"
                ],
                "activityInstanceIds": []
              }
            }
          ]
        },
        "events": [
          {
            "graph": {
              "suffix": {
  ...
```
i.e. constraint violations are included in the simulation results.

### New Behavior
Query of
```GraphQL
query NewCoolQuery {
  constraintViolations(planId: 1) {
    constraintViolations
  }
}
```
before running the simulation returns:
```JSON
{
  "data": {
    "constraintViolations": {
      "constraintViolations": {}
    }
  }
}
```

and after running the simulation returns:
```JSON
{
  "data": {
    "constraintViolations": {
      "constraintViolations": {
        "model/no_mashed": [
          {
            "windows": [
              {
                "start": 10412050000,
                "end": 86400000000
              }
            ],
            "associations": {
              "resourceIds": [
                "/fruit"
              ],
              "activityInstanceIds": []
            }
          }
        ]
      }
    }
  }
}
```

## Documentation
While not explicitly violated, should update [API docs](https://github.com/NASA-AMMOS/aerie/wiki/Aerie-GraphQL-API-Software-Interface-Specification) and [Sim Results](https://github.com/NASA-AMMOS/aerie/wiki/Simulation-Results) to mention new GraphQL query.

## Future work
Work on separating simulation result API calls contributes to possibly streaming sim results.

### CODE REVIEW

**Comment by skovati** [comment]
Working to refactor with new generic interface method introduced by [this PR](https://github.com/NASA-AMMOS/aerie/pull/191)

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/SimulationService.java [hunk 1]
```diff
@@ -3,13 +3,20 @@
...
 import gov.nasa.jpl.aerie.merlin.server.ResultsProtocol;
 import gov.nasa.jpl.aerie.merlin.server.models.PlanId;
 import org.apache.commons.lang3.tuple.Pair;
 
 public interface SimulationService {
   ResultsProtocol.State getSimulationResults(PlanId planId, RevisionData revisionData);
   Map<String, List<Pair<Duration, SerializedValue>>> getResourceSamples(PlanId planId, RevisionData revisionData);
+  Map<ActivityInstanceId, SimulatedActivity> getSimulatedActivities(PlanId planId, RevisionData revisionData);
+  Map<String, Pair<ValueSchema, List<Pair<Duration, SerializedValue>>>> getDiscreteProfiles(PlanId planId, RevisionData revisionData);
+  Map<String, List<Pair<Duration, RealDynamics>>> getRealProfiles(PlanId planId, RevisionData revisionData);
```
Comment (pcrosemurgy) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/SimulationService.java]: Looks like you may need to rebase with `develop`. Maybe you have already and it's just not pushed :)

Comment (pcrosemurgy) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/SimulationService.java]: Oops, sent too soon. Thanks!

**Comment by skovati** [comment]
Refactored to use new generic getter. Looking for additional feedback!

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java [hunk 2]
```diff
@@ -132,9 +135,10 @@ public Map<String, List<Violation>> getViolations(final PlanId planId, final Sim
 
       discreteProfiles.put(entry.getKey(), new DiscreteProfile(pieces));
     }
-
+    final var _realProfiles = this.simulationService.get(planId, revisionData,
+        r -> r.realProfiles).orElseGet(Map::of);
     final var realProfiles = new HashMap<String, LinearProfile>();
-    for (final var entry : results.realProfiles.entrySet()) {
+    for (final var entry : _realProfiles.entrySet()) {
```
Comment (pcrosemurgy) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java:150]: Very minor naming convention thing but in other areas of aerie we usually use a `$` prefix (for better or worse) instead of a `_`.

**Review by pcrosemurgy — APPROVED** [review]
Great job!

@camargo does the UI make use of constraint violations from the sim. results HTTP response? I did a quick check and I don't think it does.

**Comment by camargo** [comment]
@pcrosemurgy Yes it does (see: https://github.com/NASA-AMMOS/aerie-ui/blob/develop/src/utilities/effects.ts#L988). The UI will need to be updated to use this new action.

**Comment by pcrosemurgy** [comment]
Ohh. In that case @skovati I think we'll need to not remove the following line:
```java
.add("constraints", serializeMap(v -> serializeIterable(ResponseSerializers::serializeConstraintViolation, v), violations))
```
In a follow-up PR we can remove this once the UI has been updated to use this action.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java (hunk 2 above)
Comment (Twisol) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java:150]: Usually a `suffix$`, as I recall. The `$` is meant to indicate that it's a transparent container of something, so you might see `Optional<Foo> foo$`. It avoids either taking up a valuable name for an unwrapped `Foo foo` or belaboring the point with `optionalFoo`.

The distinction between `realProfiles` and `_realProfiles` here is somwhat different (i.e. one isn't an unwrapped version of the other), so I'm personally not fussed about switching to `$`.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/ResponseSerializers.java [hunk 3]
```diff
@@ -240,6 +239,15 @@ public static JsonValue serializeResourceSamples(final Map<String, List<Pair<Dur
         .build();
   }
 
+  public static JsonValue serializeConstraintViolations(final Map<String, List<Violation>> violations) {
+    return Json
+      .createObjectBuilder()
+      .add("constraintViolations", serializeMap(v ->
+          serializeIterable(ResponseSerializers::serializeConstraintViolation, v),
+          violations))
```
Comment (Twisol) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/ResponseSerializers.java]: It's a little hard to see from the formatting that the `v ->` on the first line only scopes the second line and not the third line. Maybe shift the `v ->` down a line so the whole lambda is in one line?

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java [hunk 4]
```diff
@@ -97,9 +98,10 @@ public Map<String, List<Violation>> getViolations(final PlanId planId, final Sim
       throw new RuntimeException("Assumption falsified -- mission model for existing plan does not exist");
     }
 
-
     final var activities = new ArrayList<ActivityInstance>();
-    for (final var entry : results.simulatedActivities.entrySet()) {
+    final var simulatedActivities = this.simulationService.get(planId, revisionData,
+        r -> r.simulatedActivities).orElseGet(Map::of);
```
Comment (Twisol) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: This formatting splits a logical method call and simultaneously groups half of it with a separate method call. Maybe something like this makes the logic flow more clearly?

```suggestion
    final var simulatedActivities = this.simulationService
        .get(planId, revisionData, r -> r.simulatedActivities)
        .orElseGet(Map::of);
```

(This happens in a few other places in this diff.)

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java (hunk 2 above)
Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java:150]: Ah sure thing, I'll push that change!

Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java:150]: Good point @Twisol, I'll stick with the `_` prefix barring any other input

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/ResponseSerializers.java (hunk 3 above)
Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/ResponseSerializers.java]: Totally agree, change incoming!

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java (hunk 4 above)
Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: Good call, latter is much clearer

Comment (Twisol) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: I have a personal preference for `Collections::emptyMap` over `Map::of`, since the latter has overloads of multiple arity and you have to recognize that the nullary overload is being used here. (My first question when I see `Map::of` is "of what?", and it takes a half-second to realize there's nothing coming in from the caller.)

No big deal though.

Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: Good point, I considered using `.orElse(Map.of())` for that same clarity reason, but `Collections::emptyMap` is even better!

Comment (Twisol) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: `Map.of()` is cheap, so `orElse` would have been fine -- but `Collections::emptyMap` exists ^_^

Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: Change has been made.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/ResponseSerializers.java (hunk 3 above)
Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/ResponseSerializers.java]: Change has been made.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java (hunk 4 above)
Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: Switched to `Collections::emptyMap`.

**Comment by skovati** [comment]
@camargo @pcrosemurgy
Branch has been updated to maintain backwards compatibility.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/MerlinBindings.java [hunk 5]
```diff
@@ -189,6 +192,23 @@ private void getResourceSamples(final Context ctx) {
       ctx.status(404).result(ResponseSerializers.serializeNoSuchPlanException(ex).toString());
     }
 }
+  private void getConstraintViolations(final Context ctx) {
+    try {
+      final var planId = parseJson(ctx.body(), hasuraPlanActionP).input().planId();
+
+      final var constraintViolations = this.simulationAction.getViolations(planId);
```
Comment (mattdailis) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/MerlinBindings.java:199]: These calls to `getViolations` and `serializeConstraintViolations` show up in an earlier commit than the definitions of those methods do, so there are a couple commits at the beginning of this PR that don't compile.

I tried reversing the order of all four commits in this PR, and they all built nicely, so if that's okay with you, I might request a reversal of the commit order.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java [hunk 6]
```diff
@@ -97,9 +99,11 @@ public Map<String, List<Violation>> getViolations(final PlanId planId, final Sim
       throw new RuntimeException("Assumption falsified -- mission model for existing plan does not exist");
     }
 
-
     final var activities = new ArrayList<ActivityInstance>();
-    for (final var entry : results.simulatedActivities.entrySet()) {
+    final var simulatedActivities = this.simulationService
+      .get(planId, revisionData, r -> r.simulatedActivities)
```
Comment (mattdailis) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: I'm noticing three calls to `simulationService.get` - any thoughts on the tradeoffs between keeping them separate, vs merging them into one call (and then pulling `simulatedActivities`, `discreteProfiles`, and `realProfiles` out afterwards)?

Comment (Twisol) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: Making one call also goes well with my note on #191 -- you can just pass `$ -> $` (or its better named equivalent, `Function.identity()`) to get an `Optional<SimulationResults>`.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/MerlinBindings.java (hunk 5 above)
Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/MerlinBindings.java:199]: Good call, I was doing some bespoke rebasing on my branch and didn't realize I ended up with invalid commits. Reversing the order works perfectly.

Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/MerlinBindings.java:199]: Change has been made.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java (hunk 6 above)
Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: I do like the idea of condensing to one function call. My first attempt at this looks like
```java
final var results = this.simulationService
    .get(planId, revisionData, Function.identity());
```
followed by
```java
final var simulatedActivities = results
    .map(r -> r.simulatedActivities)
    .orElseGet(Collections::emptyMap);
```
for each field.

Is there a simpler way of extracting a field from an `Optional` that I'm overlooking?

Comment (pcrosemurgy) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: Sorry for the churn @skovati, sloppy thinking on my part when I added `SimulationService.get()`. If you want you can change it to:
```java
Optional<SimulationResults> get(final PlanId planId, final RevisionData revisionData);
```
(Thanks @Twisol)
Or use the `Function.identity()` approach you have above.

Then in `getViolations()` I would just call that once, with something like: `final var results$ = simulationService.get(planId, revisionData);`.

Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: A-ok with either option, changing `SimulationService.get()` would require a slight change to your work for [1832](https://github.com/NASA-AMMOS/aerie/pull/191) @pcrosemurgy. Happy to make that change, just let me know your preference.

Comment (pcrosemurgy) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: No strong opinion :) Whatever you see fit is good with me

Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: I'll stick with the generic `get` function in that case, barring additional input, since otherwise we'd have two methods in `SimulationService` that both return slightly different forms of sim results.

Comment (pcrosemurgy) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: Well, they both accomplish the same thing so you'd only want 1 either way :)
```java
<T> Optional<T> get(final PlanId planId, final RevisionData revisionData, final Function<SimulationResults, T> getter);
```
or
```java
Optional<SimulationResults> get(final PlanId planId, final RevisionData revisionData);
```

So using the first form you can do:
```java
final var simResults$ = simulationService.get(planId, revisionData, Function.identity());
final var simulatedActs = simResults$.map(r -> r.simulatedActivities).orElseGet(Map::of);
```

Using the second form you can return the same data with a slightly simpler call:
```java
final var simResults$ = simulationService.get(planId, revisionData);
final var simulatedActs = simResults$.map(r -> r.simulatedActivities).orElseGet(Map::of);
...
```

**TL;DR** either is really fine, sorry for causing confusion here with my weird `get()` interface.

Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: er yeah, I was referencing the already existing `getSimulationResults` in `SimulationService`, so changing `get()` to only return `SimulationResults` felt a little redundant (of course it isn't, since the whole point is we need this optional logic for uncached sims) while the current generic implementation offers slightly more functionality

Comment (Twisol) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: > er yeah, I was referencing the already existing `getSimulationResults` in `SimulationService`, so changing `get()` to only return `SimulationResults` felt a little redundant (of course it isn't, since the whole point is we need this optional logic for uncached sims) while the current generic implementation offers slightly more functionality

It sounds like you'll have two methods that return slightly different forms of sim results no matter what your choice here is, no? If they serve different use-cases, we ought to allow them to evolve to serve those use-cases effectively. If we eventually find we can unify them, that's something we can do later. If we eventually find they're impossible to unify, we accept that we have two ways of getting sim results and that there's a fundamental difference between them.

Whether you apply your getter to the return value yourself, or let the method call your getter for you, does not seem to be materially relevant to the fact that there are two ways to get results. My vote would be to remove the getter callback, because it doesn't serve the use-case any better than using Optional's own map method does.

Comment (Twisol) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: More meaningfully, the getter argument entices you into calling get() multiple times with different getters, doing extra work to get the same initial simulation results every time. From a "pit of success" perspective, removing the getter argument avoids this enticement and makes the right thing the default.

Comment (pcrosemurgy) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: >It sounds like you'll have two methods that return slightly different forms of sim results no matter what your choice here is, no? If they serve different use-cases, we ought to allow them to evolve to serve those use-cases effectively. 

Sorry if I'm misunderstanding what you're saying but it sounds like you're saying we have the need for both forms of `get()`? 

Looking at where `SimulationService.get()` is used we only need one form (the `Optional<SimulationResults>` return form being the most ergonomic as you point out).

Comment (Twisol) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: No, I was responding to @skovati's note that get() and getSimulationResults() both get simulation results. They serve different use cases, so avoiding similarity for similarity's sake prevents them from evolving effectively to meet their respective use cases.

Comment (pcrosemurgy) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: Ahh, yeah 100% agree :)

Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: Change has been made.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java [hunk 7]
```diff
@@ -97,9 +100,13 @@ public Map<String, List<Violation>> getViolations(final PlanId planId, final Sim
       throw new RuntimeException("Assumption falsified -- mission model for existing plan does not exist");
     }
 
+    final var _results = this.simulationService.get(planId, revisionData, Function.identity());
```
Comment (Twisol) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: Now *this* is where I'd use `results$` 😉

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java (hunk 6 above)
Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: Whoops, pushed changes before I saw your input @Twisol, happy to make that change

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java (hunk 7 above)
Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: Change has been made!

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java (hunk 6 above)
Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: Alright nongeneric version has been pushed!

Comment (Twisol) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java]: Looks great! I think we've addressed Matt's initial concern here, too.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/CachedSimulationService.java [hunk 8]
```diff
@@ -33,11 +27,11 @@ public ResultsProtocol.State getSimulationResults(final PlanId planId, final Rev
   }
 
   @Override
-  public <T> Optional<T> get(final PlanId planId, final RevisionData revisionData, final Function<SimulationResults, T> getter) {
+  public Optional<SimulationResults> get(final PlanId planId, final RevisionData revisionData) {
     return this.store.lookup(planId) // Only return results that have already been cached
         .map(ResultsProtocol.ReaderRole::get)
         .map(state -> state instanceof final ResultsProtocol.State.Success s ?
-            getter.apply(s.results()) :
+            s.results() :
```
Comment (pranav-super) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/CachedSimulationService.java:34]: Why did you make this change?

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java [hunk 9]
```diff
@@ -113,9 +122,11 @@ public Map<String, List<Violation>> getViolations(final PlanId planId, final Sim
           activity.arguments(),
           Window.between(activityOffset, activityOffset.plus(activity.duration()))));
     }
-
-    final var discreteProfiles = new HashMap<String, DiscreteProfile>(results.discreteProfiles.size());
-    for (final var entry : results.discreteProfiles.entrySet()) {
+    final var _discreteProfiles = results$
```
Comment (pranav-super) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java:125]: Can you justify this change?

**Review by pranav-super — APPROVED** [review]

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/CachedSimulationService.java (hunk 8 above)
Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/CachedSimulationService.java:34]: To sum up my conversation above, I moved the getter callback outside of this `get()` function, since to get the full results object out of this getter would require a somewhat awkward
```java
final var results = simulationService
    .get(planId, revisionData, Function.identity())
    .orElseGet(Collections::emptyMap);
```
so now, instead of calling
```java
final var simActs = simulationService
    .get(planId, revisionData, r -> r.simulatedActivities)
    .orElseGet(Collections::emptyMap);
```
you can use
```java
final var simActs = simulationService
    .get(planId, revisionData)
    .map(r -> r.simulatedActivities)
    .orElseGet(Collections::emptyMap);
```
which allows easy extraction of the full `SimulationResults` object if needed.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java (hunk 9 above)
Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java:125]: Essentially, it was necessary to remove `getViolations()`'s dependence on the `SimulationResults` argument, since we want to call this method without having run the simulation, so that we can return an empty violation map if the sim was uncached but `CachedSimulationService` was being used. Instead, we are getting an `Optional<SimulationResults` object from the new getter in the `SimulationService` interface.

**Comment by skovati** [comment]
rebased onto upstream develop

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java (hunk 9 above)
Comment (mattdailis) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java:125]: One thought I have is that in maintaining backwards compatibility, we now load simulation results twice (in the case where `getViolations` is called from `GetSimulationResultsAction::run`). If we weren't planning on getting rid of that call, I would make a fuss about it, but as it stands I'm okay with leaving that in here until we remove that call in a follow-up PR :+1:

**Review by mattdailis — APPROVED** [review]

**Comment by mattdailis** [comment]
Great work, Luke! 🎉

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java (hunk 9 above)
Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java:125]: That is a good point. The backwards compatibility was a bit of an afterthought. I suppose the double loading is somewhat inherent to allowing these queries to be made separately though.

Comment (mattdailis) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java:125]: One way to "shim in" the backwards compatibility, without loading the data twice, would be to have two `getViolations`methods, one with one parameter `getViolations(PlanId)`, and the other with two parameters `getViolations(PlanId, SimulationResults)`. The first would load the results and call the second, whereas the "old" workflow would continue more or less unchanged.

If we wanted to keep both, I would push for this approach, but since we're removing one, this would add some unnecessary clutter.

Comment (skovati) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java:125]: I agree it would add some unnecessary clutter, especially since the removal of the deprecated API *should* happen within the same release (ie 0.12.1). I'll leave it as is then, barring additional feedback!

**Comment by skovati** [comment]
rebased onto upstream develop
